File: test-comprehensive-functionality.js

const { PrismaClient } = require('@prisma/client');

const prisma = new PrismaClient();

async function testComprehensiveFunctionality() {
  console.log('🧪 Starting Comprehensive Functionality Tests...\n');

  try {
    // Test 1: Database Connection
    console.log('1️⃣ Testing Database Connection...');
    await prisma.$connect();
    console.log('✅ Database connection successful\n');

    // Test 2: System Design Questions
    console.log('2️⃣ Testing System Design Questions...');
    const systemDesignQuestions = await prisma.question.findMany({
      where: { question_type: 'system_design' },
      include: { system_design_questions: true }
    });
    
    console.log(`✅ Found ${systemDesignQuestions.length} system design questions`);
    
    const categoryBreakdown = {};
    systemDesignQuestions.forEach(q => {
      categoryBreakdown[q.category] = (categoryBreakdown[q.category] || 0) + 1;
    });
    
    console.log('📊 Category breakdown:');
    Object.entries(categoryBreakdown).forEach(([cat, count]) => {
      console.log(`   ${cat}: ${count} questions`);
    });
    
    // Check for detailed answers
    const questionsWithDetails = systemDesignQuestions.filter(q => q.system_design_questions);
    console.log(`✅ ${questionsWithDetails.length} questions have detailed answers\n`);

    // Test 3: Behavioral Questions
    console.log('3️⃣ Testing Behavioral Questions...');
    const behavioralQuestions = await prisma.question.findMany({
      where: { question_type: 'behavioral' }
    });
    console.log(`✅ Found ${behavioralQuestions.length} behavioral questions\n`);

    // Test 4: Companies
    console.log('4️⃣ Testing Companies...');
    const companies = await prisma.company.findMany();
    console.log(`✅ Found ${companies.length} companies:`);
    companies.forEach(c => console.log(`   ${c.id}: ${c.name}`));
    console.log('');

    // Test 5: Users
    console.log('5️⃣ Testing Users...');
    const users = await prisma.user.findMany({
      select: { id: true, username: true, is_admin: true }
    });
    console.log(`✅ Found ${users.length} users:`);
    users.forEach(u => console.log(`   ${u.username} (${u.is_admin ? 'Admin' : 'User'})`));
    console.log('');

    // Test 6: Stories
    console.log('6️⃣ Testing Stories...');
    const stories = await prisma.story.findMany();
    console.log(`✅ Found ${stories.length} stories in database\n`);

    // Test 7: API Endpoints (basic connectivity)
    console.log('7️⃣ Testing API Endpoints...');
    
    const fetch = (await import('node-fetch')).default;
    
    const endpoints = [
      'http://localhost:3000/api/system-design-questions',
      'http://localhost:3000/api/system-design-categories',
      'http://localhost:3000/api/companies/system-design',
      'http://localhost:3000/api/companies/with-questions'
    ];
    
    for (const endpoint of endpoints) {
      try {
        const response = await fetch(endpoint);
        const data = await response.json();
        console.log(`✅ ${endpoint}: ${response.status} (${Array.isArray(data) ? data.length + ' items' : 'object'})`);
      } catch (error) {
        console.log(`❌ ${endpoint}: Failed - ${error.message}`);
      }
    }
    console.log('');

    // Test 8: Data Quality Checks
    console.log('8️⃣ Testing Data Quality...');
    
    // Check for system design questions without detailed answers
    const systemDesignWithoutDetails = await prisma.question.findMany({
      where: {
        question_type: 'system_design',
        system_design_questions: null
      }
    });
    
    if (systemDesignWithoutDetails.length === 0) {
      console.log('✅ All system design questions have detailed answers');
    } else {
      console.log(`⚠️ Found ${systemDesignWithoutDetails.length} system design questions without detailed answers`);
    }
    
    console.log('✅ Data quality checks completed');
    console.log('');

    // Test 9: Sample Data Verification
    console.log('9️⃣ Testing Sample Data...');
    
    const sampleSystemDesign = await prisma.question.findFirst({
      where: { question_type: 'system_design' },
      include: { system_design_questions: true }
    });
    
    if (sampleSystemDesign) {
      console.log(`✅ Sample system design question: "${sampleSystemDesign.question_text}"`);
      console.log(`   Category: ${sampleSystemDesign.category}`);
      console.log(`   Difficulty: ${sampleSystemDesign.difficulty}`);
      console.log(`   Tags: [${sampleSystemDesign.tags.join(', ')}]`);
      if (sampleSystemDesign.system_design_questions) {
        console.log(`   Has detailed answer: Yes`);
        console.log(`   Functional requirements: ${sampleSystemDesign.system_design_questions.functional_requirements.length}`);
      }
    }
    
    console.log('');

    // Summary
    console.log('📋 SUMMARY:');
    console.log(`✅ System Design Questions: ${systemDesignQuestions.length} (Target: 100+)`);
    console.log(`✅ Behavioral Questions: ${behavioralQuestions.length}`);
    console.log(`✅ Companies: ${companies.length}`);
    console.log(`✅ Users: ${users.length}`);
    console.log(`✅ Stories: ${stories.length}`);
    
    if (systemDesignQuestions.length >= 100) {
      console.log('🎉 SUCCESS: System design questions target achieved!');
    } else {
      console.log('⚠️ WARNING: System design questions below target');
    }
    
    console.log('\n🎯 All core functionality tests completed!');

  } catch (error) {
    console.error('❌ Test failed:', error);
  } finally {
    await prisma.$disconnect();
  }
}

testComprehensiveFunctionality();

File: test_comprehensive_functionality.py

#!/usr/bin/env python3

import requests
import json
import time
import sys

def test_api_endpoints():
    """Test all critical API endpoints"""
    base_url = "http://localhost:3000"
    
    endpoints = [
        "/api/system-design-questions",
        "/api/system-design-frameworks", 
        "/api/companies",
        "/api/questions?questionType=behavioral",
        "/api/questions?questionType=system_design"
    ]
    
    print("Testing API endpoints...")
    
    for endpoint in endpoints:
        try:
            response = requests.get(f"{base_url}{endpoint}", timeout=10)
            if response.status_code == 200:
                data = response.json()
                if isinstance(data, list):
                    count = len(data)
                elif isinstance(data, dict) and 'companies' in data:
                    count = len(data['companies'])
                else:
                    count = "N/A"
                print(f"✅ {endpoint} - Status: {response.status_code}, Count: {count}")
            else:
                print(f"❌ {endpoint} - Status: {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ {endpoint} - Error: {e}")
            return False
    
    return True

def test_authentication():
    """Test authentication flow"""
    base_url = "http://localhost:3000"
    
    print("\nTesting authentication...")
    
    # Test valid login
    try:
        response = requests.post(f"{base_url}/api/auth/login", 
                               json={"username": "admin", "password": "adminadmin"},
                               timeout=10)
        if response.status_code == 200:
            data = response.json()
            if data.get('message') == 'Login successful':
                print("✅ Valid login successful")
            else:
                print("❌ Valid login failed - unexpected response")
                return False
        else:
            print(f"❌ Valid login failed - Status: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Valid login failed - Error: {e}")
        return False
    
    # Test invalid login
    try:
        response = requests.post(f"{base_url}/api/auth/login", 
                               json={"username": "invalid", "password": "invalid"},
                               timeout=10)
        if response.status_code == 401:
            print("✅ Invalid login correctly rejected")
        else:
            print(f"❌ Invalid login not rejected - Status: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Invalid login test failed - Error: {e}")
        return False
    
    return True

def test_page_loads():
    """Test that all pages load without errors"""
    base_url = "http://localhost:3000"
    
    # Login first to get session
    session = requests.Session()
    login_response = session.post(f"{base_url}/api/auth/login", 
                                 json={"username": "admin", "password": "adminadmin"})
    
    if login_response.status_code != 200:
        print("❌ Could not login for page tests")
        return False
    
    pages = [
        "/dashboard",
        "/question-bank", 
        "/system-design-questions",
        "/story-templates",
        "/company-values",
        "/system-design-strategy",
        "/interview-strategy",
        "/progress-tracker",
        "/interview-notes",
        "/faq"
    ]
    
    print("\nTesting page loads...")
    
    for page in pages:
        try:
            response = session.get(f"{base_url}{page}", timeout=15)
            if response.status_code == 200:
                # Check for common error patterns
                content = response.text
                if "Cannot read properties of undefined" in content:
                    print(f"❌ {page} - Contains undefined property errors")
                    return False
                elif "Failed to fetch" in content:
                    print(f"❌ {page} - Contains fetch errors")
                    return False
                else:
                    print(f"✅ {page} - Loads successfully")
            else:
                print(f"❌ {page} - Status: {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ {page} - Error: {e}")
            return False
    
    return True

def test_database_connectivity():
    """Test database operations"""
    base_url = "http://localhost:3000"
    
    print("\nTesting database connectivity...")
    
    # Test data retrieval from different tables
    tests = [
        ("Questions", "/api/questions?questionType=behavioral"),
        ("System Design Questions", "/api/system-design-questions"),
        ("Companies", "/api/companies"),
        ("Frameworks", "/api/system-design-frameworks")
    ]
    
    for test_name, endpoint in tests:
        try:
            response = requests.get(f"{base_url}{endpoint}", timeout=10)
            if response.status_code == 200:
                data = response.json()
                if isinstance(data, list) and len(data) > 0:
                    print(f"✅ {test_name} - {len(data)} records retrieved")
                elif isinstance(data, dict) and data.get('companies') and len(data['companies']) > 0:
                    print(f"✅ {test_name} - {len(data['companies'])} records retrieved")
                else:
                    print(f"❌ {test_name} - No data returned")
                    return False
            else:
                print(f"❌ {test_name} - Status: {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ {test_name} - Error: {e}")
            return False
    
    return True

def check_documentation_files():
    """Check that documentation files were generated"""
    import os
    
    print("\nChecking documentation files...")
    
    required_files = [
        "behavioral-questions.md",
        "behavioral-questions.pdf", 
        "behavioral-questions-with-answers.md",
        "behavioral-questions-with-answers.pdf",
        "system-design-questions.md",
        "system-design-questions.pdf",
        "system-design-questions-with-answers.md", 
        "system-design-questions-with-answers.pdf",
        "company-values-and-strategies.md",
        "company-values-and-strategies.pdf"
    ]
    
    missing_files = []
    for file in required_files:
        if os.path.exists(file):
            size = os.path.getsize(file)
            print(f"✅ {file} - {size} bytes")
        else:
            print(f"❌ {file} - Missing")
            missing_files.append(file)
    
    return len(missing_files) == 0

def main():
    """Run comprehensive functionality tests"""
    print("🚀 Starting Comprehensive Functionality Tests\n")
    
    tests = [
        ("API Endpoints", test_api_endpoints),
        ("Authentication", test_authentication), 
        ("Database Connectivity", test_database_connectivity),
        ("Page Loads", test_page_loads),
        ("Documentation Files", check_documentation_files)
    ]
    
    results = {}
    
    for test_name, test_func in tests:
        print(f"\n{'='*50}")
        print(f"Running {test_name} Tests")
        print('='*50)
        
        try:
            result = test_func()
            results[test_name] = result
            if result:
                print(f"✅ {test_name} - PASSED")
            else:
                print(f"❌ {test_name} - FAILED")
        except Exception as e:
            print(f"❌ {test_name} - ERROR: {e}")
            results[test_name] = False
    
    # Summary
    print(f"\n{'='*50}")
    print("TEST SUMMARY")
    print('='*50)
    
    passed = sum(1 for result in results.values() if result)
    total = len(results)
    
    for test_name, result in results.items():
        status = "PASSED" if result else "FAILED"
        emoji = "✅" if result else "❌"
        print(f"{emoji} {test_name}: {status}")
    
    print(f"\nOverall: {passed}/{total} tests passed")
    
    if passed == total:
        print("🎉 ALL TESTS PASSED! The application is fully functional.")
        return 0
    else:
        print("⚠️  Some tests failed. Please review the issues above.")
        return 1

if __name__ == "__main__":
    exit_code = main()
    sys.exit(exit_code)
